spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -380,40 +438,87 @@ public boolean isLongLived() {
 
 		@Override
 		public void run() {
-			this.consumerThread = Thread.currentThread();
 			this.count = 0;
 			this.last = System.currentTimeMillis();
 			if (isRunning() && this.definedPartitions != null) {
 				initPartitionsIfNeeded();
 			}
+			ConsumerRecords<K, V> currentRecords = null;
+			Iterator<ConsumerRecord<K, V>> currentRecordsIterator = this.emptyIterator;
+			ConsumerRecord<K, V> currentRecord = null;
 			while (isRunning()) {
 				try {
 					if (this.logger.isTraceEnabled()) {
-						this.logger.trace("Polling...");
+						this.logger.trace("Polling (paused=" + this.paused + ")...");
 					}
-					ConsumerRecords<K, V> records = this.consumer.poll(getPollTimeout());
-					if (records != null) {
+					if (this.paused || !currentRecordsIterator.hasNext()) {
+						ConsumerRecords<K, V> records = this.consumer.poll(getPollTimeout());
 						if (this.logger.isDebugEnabled()) {
 							this.logger.debug("Received: " + records.count() + " records");
 						}
-						Iterator<ConsumerRecord<K, V>> iterator = records.iterator();
-						while (iterator.hasNext()) {
-							final ConsumerRecord<K, V> record = iterator.next();
+						if (!this.paused) {
+							currentRecords = records;
+							currentRecordsIterator = records != null ? records.iterator() : this.emptyIterator;
+						}
+					}
+					if (this.paused && !this.pausedForException && !this.listenerHasReturned) {
+						this.logger.trace("Paused consumer exited poll and listener has still not returned");
+						continue;
```

Comment:
I have just reworked it to pass the batch instead of individual records and it's much simpler.

Still using a Future for now, but could easily convert to a queue so we always have work "in hand" for the listener to process rather than polling only when the listener handles the current batch - good idea.